src/state_flow/core.clj

```
@@ -105,35 +105,56 @@
+    (on-error pair)))
+
 (defn run!
   "Like run, but prints a log and throws an error when the flow fails with an exception"
   ([flow]
    (run! flow {}))
   ([flow initial-state]
    (let [pair (run flow initial-state)]
-     (when (e/failure? (first pair))
-       (let [description (->> pair second description-stack format-description)
-             message (str "Flow " "\"" description "\"" " failed with exception")]
-         (log/info (m/extract (first pair)) message)
-         (throw (ex-info message {}))))
-     pair)))
+     (or (run-policy-on-error! pair log-and-throw-error!)
+         pair))))
 
 (defn run*
   "Run a flow with specified parameters
 
   Receives optional parameter maps
-  `init`, a function with no arguments that returns the initial state.
+  `init`, a function with no arguments that returns the initial state
   `cleanup`, function receiving the final state to perform cleanup if necessary
-  `runner`, function that will receive a flow and an initial state and execute the flow"
-  [{:keys [init cleanup runner]
-    :or   {init    (constantly {})
-           cleanup identity
-           runner  run!}}
+  `runner`, function that will receive a flow and an initial state and execute the flow
+  `on-error`, funtion that, when a flow results in an error, will receive the final result pair. Defaults to `log-and-throw-error!`"
+  [{:keys [init cleanup runner on-error]
+    :or   {init     (constantly {})
+           cleanup  identity
+           runner   run
+           on-error log-and-throw-error!}}
    flow]
-  (let [initial-state              (init)
-        [_ final-state :as result] (runner flow initial-state)]
-    (cleanup final-state)
-    result))
+  (let [initial-state (init)
+        pair          (runner flow initial-state)]
+    (try
+      (cleanup (second pair))
+      (catch java.lang.Throwable t
+        (log/error t "Error when trying to cleanup after exception")))
+    (or (run-policy-on-error! pair on-error)
+        pair)))
```

if `on-error` returns `nil` then `pair` will be returned, which might be considered unexpected

---

It's a new behavior, so it won't be a breaking change anyways.

About being considered unexpected: that's true, but I also think throwing information away is not a good practice. And if an exception is ignored, I don't see a reason not to return the result.